File: adapters/ui-swift/Sources/FluxUIKit/Props.swift
//  Props.swift
//  FluxUIKit — the prop map contract (Appendix C §C.1 `Props`).

/// A flat map of `(prop_idx, value)` pairs attached to an IR node.
///
/// Mirrors `flux_syntax::Props`: a content-addressed bag of fields. Field
/// access is by `PropIdx` (O(1)). `hash` is a stable content hash used by the
/// runtime's reconciler to skip referentially-equal subtrees.
public struct Props: Sendable, Hashable {
    /// The underlying field map.
    public let fields: [PropIdx: FluxValue]
    private let digest: UInt64

    /// Build a prop map. The content hash is computed up front so equality and
    /// lookups are O(1) later.
    public init(_ fields: [PropIdx: FluxValue] = [:]) {
        self.fields = fields
        var hasher = Hasher()
        for index in fields.keys.sorted() {
            hasher.combine(index)
            // `fields` is a dictionary; the key is guaranteed present.
            hasher.combine(fields[index]!)
        }
        self.digest = UInt64(bitPattern: Int64(hasher.finalize()))
    }

    /// The content hash (BLAKE3 in the Rust IR; a portable hash here).
    public var hash: UInt64 { digest }

    /// Look up a raw value by prop index.
    public func get(_ index: PropIdx) -> FluxValue? { fields[index] }

    /// Resolve a string prop.
    public func getString(_ index: PropIdx) -> String? {
        if case .str(let s) = fields[index] ?? .null { s } else { nil }
    }

    /// Resolve an integer prop.
    public func getInt(_ index: PropIdx) -> Int64? {
        if case .int(let i) = fields[index] ?? .null { i } else { nil }
    }

    /// Resolve a float prop.
    public func getFloat(_ index: PropIdx) -> Double? {
        if case .float(let f) = fields[index] ?? .null { f } else { nil }
    }

    /// Resolve a boolean prop.
    public func getBool(_ index: PropIdx) -> Bool? {
        if case .bool(let b) = fields[index] ?? .null { b } else { nil }
    }

    /// Resolve a handler-reference prop.
    public func getHandler(_ index: PropIdx) -> FluxHandlerId? {
        if case .handlerRef(let h) = fields[index] ?? .null { h } else { nil }
    }

    /// Resolve a record prop (itself a `Props`).
    public func getRecord(_ index: PropIdx) -> Props? {
        if case .record(let p) = fields[index] ?? .null { p } else { nil }
    }

    /// Resolve a color prop (see `FluxColor`).
    public func getColor(_ index: PropIdx) -> FluxColor? {
        getRecord(index).flatMap(FluxColor.init(record:))
    }

    /// Resolve a font prop (see `FluxFount`).
    public func getFont(_ index: PropIdx) -> FluxFount? {
        getRecord(index).flatMap(FluxFount.init(record:))
    }

    public static func == (lhs: Props, rhs: Props) -> Bool { lhs.fields == rhs.fields }

    public func hash(into hasher: inout Hasher) { hasher.combine(digest) }
}

File: adapters/ui-kotlin/src/main/kotlin/dev/flux/ui/Props.kt
package dev.flux.ui

import kotlin.collections.List as KList

/**
 * A flat, ordered map of component-local prop indices to decoded values.
 *
 * Mirrors `flux_ir::Props` (Appendix C.1): a `Vec<(PropIdx, Value)>` keyed by a
 * per-component `u16` field index. Adapters read props through the typed
 * accessors ([getString], [getBool], [getRecord], ...) rather than by raw
 * index, so missing/renamed fields degrade to `null`/default instead of
 * throwing. The contract a dev adapter binds against is exactly these accessors
 * plus the record layouts documented in Appendix F.
 */
public data class Props(
    val fields: KList<Props.Field>,
) {
    /**
     * A single prop field: a component-local index and its decoded value.
     * Field order within a [Props] is unspecified and must not be relied on.
     */
    public data class Field(
        val index: UShort,
        val value: FluxValue,
    )

    /** Returns the [FluxValue] stored at [index], or `null` if absent. */
    public fun get(index: UShort): FluxValue? = fields.firstOrNull { it.index == index }?.value

    /** String value at [index], or `null` when absent or not a string. */
    public fun getString(index: UShort): String? = (get(index) as? FluxValue.Str)?.value

    /** Integer value at [index], or `null` when absent or not an integer. */
    public fun getInt(index: UShort): kotlin.Long? = (get(index) as? FluxValue.Int)?.value

    /** Float value at [index], or `null` when absent or not a float. */
    public fun getFloat(index: UShort): kotlin.Double? = (get(index) as? FluxValue.Float)?.value

    /** Boolean value at [index], falling back to [default] when absent/mistyped. */
    public fun getBool(
        index: UShort,
        default: Boolean,
    ): Boolean = (get(index) as? FluxValue.Bool)?.value ?: default

    /** Record value at [index], or `null` when absent or not a record. */
    public fun getRecord(index: UShort): FluxValue.Record? = get(index) as? FluxValue.Record

    /** List value at [index], or `null` when absent or not a list. */
    public fun getList(index: UShort): KList<FluxValue>? = (get(index) as? FluxValue.List)?.items

    /**
     * Handler id at [index]. Returns `0` when the prop is absent or not a
     * handler reference — `0` is the reserved "no handler" id in the IR.
     */
    public fun getHandler(index: UShort): UInt = (get(index) as? FluxValue.HandlerRef)?.handlerId ?: 0u

    // Audit D5: the server lowers Color/Font records POSITIONALLY (fields
    // 0..3 / 0..2). The previous FNV-name lookups never matched, so colors
    // and fonts were silently dropped on Android. Field order mirrors
    // adapters/ui-swift Color.swift + TextAdapter.swift.

    /** Float value at the record's positional [slot], or `null`. */
    private fun FluxValue.Record.floatAt(slot: Int): kotlin.Double? =
        (fields.getOrNull(slot)?.value as? FluxValue.Float)?.value

    private fun FluxValue.Record.stringAt(slot: Int): String? {
        return (fields.getOrNull(slot)?.value as? FluxValue.Str)?.value
    }

    /** Int value at the record's positional [slot], or `null`. (audit D6) */
    private fun FluxValue.Record.intAt(slot: Int): kotlin.Long? =
        (fields.getOrNull(slot)?.value as? FluxValue.Int)?.value


    /** Decodes the `Color` record at [index] into a [FluxColor], or `null`. */
    public fun getColor(index: UShort): FluxColor? {
        val record = getRecord(index) ?: return null
        val red = record.floatAt(0) ?: return null
        val green = record.floatAt(1) ?: return null
        val blue = record.floatAt(2) ?: return null
        val alpha = record.floatAt(3) ?: 1.0
        return FluxColor(red, green, blue, alpha)
    }

    /** Decodes the `Font` record at [index] into a [FluxFont], or `null`. */
    public fun getFont(index: UShort): FluxFont? {
        val record = getRecord(index) ?: return null
        val size = record.floatAt(0) ?: return null
        return FluxFont(
            size = size,
            weight = record.stringAt(1),
            family = record.stringAt(2),
        )
    }

    public companion object {
        /** An empty prop bag — used when a node carries no props. */
        public val EMPTY: Props = Props(emptyList())
    }
}
